src/ucp/wireup/wireup_cm.c

```diff
@@ -229,37 +227,226 @@ static void ucp_cm_client_connect_cb(uct_ep_h ep, void *arg,
+{
+    ucp_ep_h ucp_ep            = (ucp_ep_h)arg;
+    ucp_worker_h worker        = ucp_ep->worker;
+    uct_worker_cb_id_t prog_id = UCS_CALLBACKQ_ID_NULL;
+    ucp_cm_client_connect_progress_arg_t *progress_arg;
+
+    if (status != UCS_OK) {
+        goto err_out;
+    }
+
+    if (!ucs_test_all_flags(remote_data->field_mask,
+                            UCT_CM_REMOTE_DATA_FIELD_DEV_ADDR        |
+                            UCT_CM_REMOTE_DATA_FIELD_DEV_ADDR_LENGTH |
+                            UCT_CM_REMOTE_DATA_FIELD_CONN_PRIV_DATA  |
+                            UCT_CM_REMOTE_DATA_FIELD_CONN_PRIV_DATA_LENGTH)) {
+        status = UCS_ERR_UNSUPPORTED;
+        goto err_out;
+    }
+
+    progress_arg = ucs_malloc(sizeof(*progress_arg),
+                              "ucp_cm_client_connect_progress_arg_t");
+    if (progress_arg == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto err_out;
+    }
+
+    progress_arg->sa_data = ucs_malloc(remote_data->conn_priv_data_length,
+                                        "sa data");
+    if (progress_arg->sa_data == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto err_free_arg;
+    }
+
+    progress_arg->dev_addr = ucs_malloc(remote_data->dev_addr_length,
+                                        "device address");
+    if (progress_arg->dev_addr == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto err_free_sa_data;
+    }
+
+    progress_arg->ucp_ep    = ucp_ep;
+    progress_arg->uct_cm_ep = uct_cm_ep;
+    memcpy(progress_arg->dev_addr, remote_data->dev_addr,
+           remote_data->dev_addr_length);
+    memcpy(progress_arg->sa_data, remote_data->conn_priv_data,
+           remote_data->conn_priv_data_length);
+
+    uct_worker_progress_register_safe(worker->uct,
+                                      ucp_cm_client_connect_progress,
+                                      progress_arg, UCS_CALLBACKQ_FLAG_ONESHOT,
+                                      &prog_id);
+    ucp_worker_signal_internal(ucp_ep->worker);
+    return;
+
+err_free_sa_data:
+    ucs_free(progress_arg->sa_data);
+err_free_arg:
+    ucs_free(progress_arg);
+err_out:
+    ucp_ep->flags &= ~UCP_EP_FLAG_LOCAL_CONNECTED;
+    ucp_worker_set_ep_failed(worker, ucp_ep, uct_cm_ep,
+                             ucp_ep_get_cm_lane(ucp_ep), status);
 }
 
-static void ucp_cm_disconnect_cb(uct_ep_h ep, void *arg)
+/*
+ * Internal flush completion callback which is a part of close protocol,
+ * this flush was initiated by remote peer in disconnect callback on CM lane.
+ */
+static void ucp_ep_cm_disconnect_flushed_cb(ucp_request_t *req)
 {
-    ucs_error("UCP close protocol is not completely implemented");
+    ucp_ep_h ucp_ep = req->send.ep;
+
+    UCS_ASYNC_BLOCK(&ucp_ep->worker->async);
+    ucp_ep_cm_disconnect_cm_lane(ucp_ep);
+    ucp_request_complete_send(req, UCS_OK);
+    UCS_ASYNC_UNBLOCK(&ucp_ep->worker->async);
+}
+
+static unsigned ucp_ep_cm_remote_disconnect_progress(void *arg)
```

Comment:
- param could be ucp_ep_h
- rename to ucp_ep_cm_do_disconnect

Reply:
what about https://github.com/openucx/ucx/pull/4445#discussion_r349912367 ?